pkg/controller/external_nofork.go

```diff
@@ -264,18 +278,53 @@ func (c *NoForkConnector) Connect(ctx context.Context, mg xpresource.Managed) (m
 	}
 
 	return &noForkExternal{
-		ts:             ts,
-		resourceSchema: c.config.TerraformResource,
-		config:         c.config,
-		params:         params,
-		rawConfig:      rawConfig,
-		logger:         logger,
-		metricRecorder: c.metricRecorder,
-		opTracker:      opTracker,
+		ts:                             ts,
+		resourceSchema:                 c.config.TerraformResource,
+		config:                         c.config,
+		params:                         params,
+		initProviderExclusiveParamKeys: initProviderExclusiveParams,
+		rawConfig:                      rawConfig,
+		logger:                         logger,
+		metricRecorder:                 c.metricRecorder,
+		opTracker:                      opTracker,
 	}, nil
 }
 
-func (n *noForkExternal) getResourceDataDiff(ctx context.Context, s *tf.InstanceState) (*tf.InstanceDiff, error) {
+func deleteInstanceDiffAttribute(instanceDiff *tf.InstanceDiff, paramKey string) error {
+	delete(instanceDiff.Attributes, paramKey)
+
+	keyComponents := strings.Split(paramKey, ".")
+	if len(keyComponents) < 2 {
+		return nil
+	}
+
+	keyComponents[len(keyComponents)-1] = "%"
+	lengthKey := strings.Join(keyComponents, ".")
+	if lengthValue, ok := instanceDiff.Attributes[lengthKey]; ok {
+		newValue, err := strconv.Atoi(lengthValue.New)
+		if err != nil {
+			return errors.Wrap(err, "cannot convert instance diff attribute to integer")
+		}
+
+		// TODO: consider what happens if oldValue = ""
+		oldValue, err := strconv.Atoi(lengthValue.Old)
+		if err != nil {
+			return errors.Wrap(err, "cannot convert instance diff attribute to integer")
+		}
+
+		newValue -= 1
+		if oldValue == newValue {
+			delete(instanceDiff.Attributes, lengthKey)
```

Comment:
Looks like we don't need to handle the length attribute (at least for maps). We can also check if this is actually the case for lists. We may just drop handling the length key hopefully.

Reply:
Dropping the length attribute for maps and sets is safe. For lists, doing so doesn't work — diff doesn't apply correctly, probably because list is not allocated with correct length. To be investigated later.